module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,204 @@
+	fulfilledHeights      chan uint64
+	thresholdChan         chan uint64
+	heightRangeTargetChan chan uint64
+
+	lastFulfilledHeight uint64
+	lastPrunedHeight    uint64
+
+	// the height range is the range of heights between the last pruned and last fulfilled
+	// heightRangeTarget is the target minimum value for this range, so that after pruning
+	// the height range is equal to the target.
+	heightRangeTarget uint64
+
+	// threshold defines the maximum height range and how frequently pruning is performed.
+	// once the height range reaches `heightRangeTarget+threshold`, `threshold` many blocks
+	// are pruned
+	threshold uint64
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataPrunerMetrics
+
+	component.Component
+	cm *component.ComponentManager
+}
+
+type PrunerOption func(*Pruner)
+
+// WithHeightRangeTarget is used to configure the pruner with a custom
+// height range target.
+func WithHeightRangeTarget(heightRangeTarget uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.heightRangeTarget = heightRangeTarget
+	}
+}
+
+// WithThreshold is used to configure the pruner with a custom threshold.
+func WithThreshold(threshold uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.threshold = threshold
+	}
+}
+
+func WithPruneCallback(callback func(context.Context) error) PrunerOption {
+	return func(p *Pruner) {
+		p.pruneCallback = callback
+	}
+}
+
+// NewPruner creates a new Pruner.
+func NewPruner(logger zerolog.Logger, metrics module.ExecutionDataPrunerMetrics, storage tracker.Storage, opts ...PrunerOption) (*Pruner, error) {
+	lastPrunedHeight, err := storage.GetPrunedHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pruned height: %w", err)
+	}
+
+	fulfilledHeight, err := storage.GetFulfilledHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulfilled height: %w", err)
+	}
+
+	fulfilledHeights := make(chan uint64, 32)
+	fulfilledHeights <- fulfilledHeight
+
+	p := &Pruner{
+		logger:                logger.With().Str("component", "execution_data_pruner").Logger(),
+		storage:               storage,
+		pruneCallback:         func(ctx context.Context) error { return nil },
+		fulfilledHeights:      fulfilledHeights,
+		thresholdChan:         make(chan uint64),
+		heightRangeTargetChan: make(chan uint64),
+		lastFulfilledHeight:   fulfilledHeight,
+		lastPrunedHeight:      lastPrunedHeight,
+		heightRangeTarget:     defaultHeightRangeTarget,
+		threshold:             defaultThreshold,
+		metrics:               metrics,
+	}
+	p.cm = component.NewComponentManagerBuilder().
+		AddWorker(p.loop).
+		Build()
+	p.Component = p.cm
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p, nil
+}
+
+// NotifyFulfilledHeight notifies the Pruner of the latest fulfilled height.
+func (p *Pruner) NotifyFulfilledHeight(height uint64) {
+	if util.CheckClosed(p.cm.ShutdownSignal()) {
+		return
+	}
+
+	p.fulfilledHeights <- height
```

Comment:
I think this block execution if the pruner falls behind. We can just skip the notification if the buffer is full since we only care about the highest height.